### ISSUE TITLE
Generics with IDisposable type parameters

### ISSUE THREAD

**Issue by DaveInCaz** [DotNetAnalyzers/IDisposableAnalyzers#271]
There is a case I noted didn't get flagged with a warning but I'm a little unsure if it should have...

These two cases are treated differently:

```
    class Test
    {
        public DisposeMe GetOne() { return new DisposeMe(); }

        public IEnumerable<DisposeMe> GetMany() { yield return new DisposeMe(); }


        void TestOne()
        {
            // This is flagged with IDISP001
            var one = GetOne();
        }

        void TestMany()
        {
            // This is not flagged - but the contents of many DO require disposal
            var many = GetMany();
        }
    }
```

(where `DisposeMe` is `IDisposable`).

It seems to me that these two cases probably should be handled the same way? I'm not 100% certain of that, but if you imagine that instead of returning `IEnumerable<DisposeMe>` it returned a container object which was *itself* `IDisposable` then they would be treated identically already. So maybe its a quirk of using generics where the generic parameter is `IDisposable` that we get different behavior. Because normally an object which contained an `IDisposable` member would be `IDisposable` itself.

A potential "rule" would be to flag any generic object where a type parameter is `IDisposable` as though it was `IDisposable` itself. But maybe that is too strict, because such a generic is not guaranteed to have an `IDisposable` member just because it uses an  `IDisposable` for ... something.

Perhaps just looking for the standard collection types would be better?

**Comment by JohanLarsson** [comment]
There is not much support for collections, analysis gets pretty tricky.

```xs
// Keep track of what this index was previously assigned with and if it was disposed etc.
xs[0] = new Disposable();
```

As it is now we take a very simplistic approach.

**Comment by DaveInCaz** [comment]
One approach might be to look for "low hanging fruit" - any simple, straightforward cases, and identify those instances with warnings. I'll try to give some more thought to what those might be (if any).

If I had some collection that might or might not contain `IDisposables` I would probably err on the side of something like this...

```
Dispose()
{
...
   foreach(var item in items.OfType<IDisposable>())
   {
      item?.Dispose();
   }
...
}

```

but that assumes the context is such that `items` as a whole is owned by this class and its proper to dispose of its contents in the first place! That seems like the tricky part.

**Comment by JohanLarsson** [comment]
Yes, I've been taking that approach only adding support for some cases that have come up.
